### DISCUSSION TITLE
Welcome to cocotbext-axi Discussions!

### DISCUSSION THREAD

**Discussion by alexforencich** [alexforencich/cocotbext-axi#23] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by mvsoliveira** [comment]
Hi @alexforencich, thanks for starting this discussion. Many thanks too for creating cocotbext-axi. I am using it for a project at CERN and serves very well the simulation of a chip2chip interface. 

I was wondering how can I reuse the same software to drive my test benches using your cocotbext-axi and the actual hardware running in the Ultrascale Zynq MPSoC.

Please let me know if you can share any examples or tips here or in the other discussion I started at:
https://github.com/cocotb/cocotb/discussions/3615

**Reply by alexforencich** [reply]
Unfortunately cocotb isn't directly compatible with python's asyncio, so there isn't really a nice way to do that. My recommendation is to split your python code so you can have a pluggable interface implementation, and then you can use the cocotb one for cocotb and a different one for interfacing with your hardware. Simple way to do this is to have an object that represents the hardware that has read and write methods, and then create two versions of this object, one for the simulator and one for the actual hardware.

Although it's entirely possible even doing something like that isn't going to be completely sufficient; another option maybe is to insert a TCP socket or some other form of IPC that can be used to interface your code with cocotb. So with that setup, you would have one interface module for the hardware, and a different interface module that uses IPC to talk to the simulator.

Someone is currently looking in to how to connect cocotbext-pcie to qemu, and currently they're using a TCP socket to make that connection. I guess that's not quite the same thing as what you're asking about, but is a data point.
